Fix `alwaysUseBid` bid setting

kmjennison opened the pull request:
Fix for #389.

protonate commented on src/prebid.js:
```diff
@@ -171,6 +171,36 @@ function getWinningBidTargeting() {
   return winners;
 }
 
+exports.getWinningBidTargeting = function() {
```
It seems you are exporting functions for the sole purpose of making them available to test. We've been following the general principle that we should test the public API of a module, that is the `exports` and ES2015 style named `export`s, but not the internal implementation of a module. Test coverage should show that comprehensive testing of the public API also covers internal implementation, or where it does not it indicates insufficient tests or cruft in the code. In any event, the export of functions in order to test should be avoided as it can be misleading as to the intent of the module.

kmjennison replied:
This was a helpful description of the test philosophy, thank you.

protonate commented on src/prebid.js:
```diff
@@ -171,6 +171,36 @@ function getWinningBidTargeting() {
+          
+          // Get only the non-standard keys of the losing bids, since we
+          // don't want to override the standard keys of the winning bid.
+          if (standardKeys.indexOf(key) > -1) {
```
could use `standardKeys.includes(key)` here

DubKF replied:
regarding standardKeys.includes - this gave me an error in IE while the previously proposed solution works fine. Therefore, either a polyfill might be needed or indexOf should be left untouched [Stackoverflow covers this as well](http://stackoverflow.com/questions/31119300/ie11-object-doesnt-support-property-or-method-includes-javascript-window)

protonate replied:
Ok, let's leave it as is and we'll see about adding a polyfill for `Array.prototype.includes` later.

protonate commented on src/prebid.js:
```diff
@@ -655,4 +698,16 @@ pbjs.enableSendAllBids = function () {
   pb_sendAllBids = true;
 };
 
+pbjs.disableSendAllBids = function () {
```
The spec for `enableSendAllBids` was as a "one way" config as it seemed there was no case where turning this off was needed once it was turned on. Are you aware of such of a case?

kmjennison replied:
You're right—I do not see a case where it would be useful to disable sending all bids. I've removed this public function in the latest pull requests, instead using a variable on `pbjs` so that we can clean up calls to `enableSendAllBids` between tests.

protonate commented on test/spec/prebid_spec.js:
```diff
@@ -0,0 +1,200 @@
+var assert = require("assert");
```
If possible it would be great to move these tests to `test/spec/unit/pbjs_api_spec.js` by testing the public `setTargetingForGPTAsync` function with settings that would cover each of the additional conditions introduced by this fix.

protonate commented:
Thanks for contributing this fix, looks good, please see the comments about test strategy and remove unneeded exports, move tests to public API.

protonate commented:
We would like to include this in our pending release. We'll go ahead and merge as-is unless you have an update in the works.

kmjennison commented:
Thanks for the comments. I'll have a PR addressing them within the next few hours.

kmjennison commented:
@protonate These latest commits address your comments:

- moved testing to the public `setTargetingForGPTAsync` rather than testing the internal implementation
- removed unneeded exports from prebid.js

I also improved a few existing tests for `setTargetingForGPTAsync` to verify the targeting key/values are correct.

mkendall07 commented:
@kmjennison 
Can you resolve conflicts? Thanks

protonate commented:
@mkendall07 @kmjennison The conflict is due to Deal ID PR (merged) and this PR touching the same code so I've resolved conflicts, I'm going to get a review from @matthewlane to confirm Deal ID functionality, test and then merge this PR.

protonate commented:
This has been merged manually, closing, thanks for the fix @kmjennison.
